File: practice/algorithm/queen.java
/*
 * 代码完成时间： 2014-12-19 23:51 AM
 * 单数组方式实现八皇后问题
 * java语言实现
 */


/**
 *
 * @author dainslef
 */

public class queen {

	public static void main(String[] args) {
		GetQueen q = new GetQueen();
		q.queen(0);
		System.out.println("\n共找到了" + q.getCount() + "种解法。");
	}

}

class GetQueen
{
	private final int QUEEN_SIZE = 8;
	private int count = 0;
	private int[] chessBoard = new int[QUEEN_SIZE];

	int getCount() {
		return count;
	}

	void queen(int row) {
		if (row == QUEEN_SIZE) {
			count++;
			System.out.println("\n找到了第" + count + "种方法：");
			show();
		}
		else {
			for (int i = 0; i < QUEEN_SIZE; i++)
				if (find(row, i))
				{
					chessBoard[row] = i;
					queen(row + 1);
				}
		}
	}

	private boolean find(int row, int line) {
		for (int i = 0; i < row; i++)
			if (line == chessBoard[i] || Math.abs(chessBoard[i] - line) == Math.abs(i - row))
				return false;
		return true;
	}

	private void show() {
		for (int i = 0; i < QUEEN_SIZE; i++)
			for (int p = 0; p < QUEEN_SIZE; p++) {
				if (p == chessBoard[i])
					System.out.print("X  ");
				else
					System.out.print("_  ");
				if (p == QUEEN_SIZE - 1)
					System.out.println();
			}
	}
}

File: practice/algorithm/queen.scala
/*
 * 代码完成时间： 2016-6-17 17:29
 * 单数组方式实现八皇后问题
 * Scala语言版本，单数组Lambda内部递归实现
 */


/**
 *
 * @author dainslef
 */

object Queen extends App {

  implicit class Self(in: Self => Int => Unit) {
    def apply(self: Self) = in(self)
  }

  val QUEEN_SIZE = 8
  val chessBoard = new Array[Int](QUEEN_SIZE)

  var count = 0

  import java.lang.Math.abs

  // 使用Lambda内部自递归
  ((x: Self) => x(x)) ((queen: Self) =>
    (row: Int) =>
      if (row == QUEEN_SIZE) {
        count += 1
        println(s"\n找到了第 $count 种方法：")
        for {
          row <- 0 until QUEEN_SIZE
          line <- 0 until QUEEN_SIZE
        } {
          print(s"${if (line == chessBoard(row)) 'X' else '_'} ")
          if (line == QUEEN_SIZE - 1) println
        }
      } else
        for {
          line <- 0 until QUEEN_SIZE
          if 0 until row filter { i =>
            chessBoard(i) == line || abs(chessBoard(i) - line) == abs(i - row)
          } isEmpty
        } {
          chessBoard(row) = line
          queen(queen)(row + 1)
        }
  )(0)

  println(s"\n共找到了 $count 种解法。")

}